File: Sources/XMTP/AuthorizedIdentity.swift
//
//  AuthorizedIdentity.swift
//
//
//  Created by Pat Nakajima on 11/17/22.
//

import Foundation

struct AuthorizedIdentity {
	var address: String
	var authorized: PublicKey
	var identity: PrivateKey

	func createAuthToken() async throws -> String {
		let authData = AuthData(walletAddress: address)
		let authDataBytes = try authData.serializedData()
		let signature = try await identity.sign(Util.keccak256(authDataBytes))

		var token = Token()

		token.identityKey = authorized
		token.authDataBytes = authDataBytes
		token.authDataSignature = signature

		return try token.serializedData().base64EncodedString()
	}

	var toBundle: PrivateKeyBundle {
		get throws {
			var bundle = PrivateKeyBundle()
			let identity = identity
			let authorized = authorized

			bundle.v1.identityKey = identity
			bundle.v1.identityKey.publicKey = authorized
			return bundle
		}
	}
}

// In an extension so we don't lose the normal struct init()
extension AuthorizedIdentity {
	init(privateKeyBundleV1: PrivateKeyBundleV1) {
		address = privateKeyBundleV1.identityKey.walletAddress
		authorized = privateKeyBundleV1.identityKey.publicKey
		identity = privateKeyBundleV1.identityKey
	}
}

File: library/src/main/java/org/xmtp/android/library/AuthorizedIdentity.kt
package org.xmtp.android.library

import android.util.Base64
import com.google.crypto.tink.subtle.Base64.encodeToString
import org.xmtp.android.library.messages.AuthDataBuilder
import org.xmtp.android.library.messages.PrivateKey
import org.xmtp.android.library.messages.PrivateKeyBuilder
import org.xmtp.android.library.messages.PrivateKeyBundle
import org.xmtp.android.library.messages.PrivateKeyBundleV1
import org.xmtp.android.library.messages.PublicKey
import org.xmtp.android.library.messages.Token
import org.xmtp.android.library.messages.walletAddress
import org.xmtp.proto.message.contents.PrivateKeyOuterClass

data class AuthorizedIdentity(
    var address: String,
    var authorized: PublicKey,
    var identity: PrivateKey,
) {

    constructor(privateKeyBundleV1: PrivateKeyBundleV1) : this(
        privateKeyBundleV1.identityKey.walletAddress,
        privateKeyBundleV1.identityKey.publicKey,
        privateKeyBundleV1.identityKey,
    )

    fun createAuthToken(): String {
        val authData = AuthDataBuilder.buildFromWalletAddress(walletAddress = address)
        val signature = PrivateKeyBuilder(identity).sign(Util.keccak256(authData.toByteArray()))

        val token = Token.newBuilder().also {
            it.identityKey = authorized
            it.authDataBytes = authData.toByteString()
            it.authDataSignature = signature
        }.build().toByteArray()
        return encodeToString(token, Base64.NO_WRAP)
    }

    val toBundle: PrivateKeyBundle
        get() {
            return PrivateKeyOuterClass.PrivateKeyBundle.newBuilder().also {
                it.v1Builder.also { v1Builder ->
                    v1Builder.identityKey = identity
                    v1Builder.identityKeyBuilder.publicKey = authorized
                }.build()
            }.build()
        }
}
